From 344fa57c4a0b44ae9fc5a0f769393d314936506a Mon Sep 17 00:00:00 2001
From: Sarika Sharma <quic_sarishar@quicinc.com>
Date: Mon, 10 Feb 2025 09:18:36 +0530
Subject: [PATCH] wifi: ath11k: fix slab out of bound issue for tx_stats
 mu_group

Currently array tx_stats->mu_group size is MAX_MU_GROUP_ID(64).
However, when assigning values to the array, the iteration runs from 0 to
MAX_MU_GROUP_ID(64).
This could potentially cause array out of bounds issue and lead to below
call trace:

dump_backtrace+0xd4/0xf4
show_stack+0x14/0x1c
dump_stack_lvl+0x44/0x58
print_report+0xa0/0x4ac
kasan_report+0xac/0xbc
__asan_load8+0x24/0x84
ath11k_debugfs_sta_add_tx_stats+0x9c4/0xea8 [ath11k]
ath11k_dp_tx_htt_monitor_mode_ring_config+0x124c/0x49c4 [ath11k]
ath11k_dp_htt_htc_t2h_msg_handler+0x6d8/0xcf0 [ath11k]
ath11k_htc_rx_completion_handler+0x69c/0x6b4 [ath11k]
ath11k_ce_per_engine_service+0x360/0x564 [ath11k]

Hence, fix issue to iterate while assigning values to array only up to
MAX_MU_GROUP_ID(64)-1.

patch-dependency: 069-ath11k-add-HE-stats-in-peer-stats.patch

Signed-off-by: Sarika Sharma <quic_sarishar@quicinc.com>
---
 drivers/net/wireless/ath/ath11k/debugfs_sta.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/net/wireless/ath/ath11k/debugfs_sta.c b/drivers/net/wireless/ath/ath11k/debugfs_sta.c
index 8838e32..3abe796 100644
--- a/drivers/net/wireless/ath/ath11k/debugfs_sta.c
+++ b/drivers/net/wireless/ath/ath11k/debugfs_sta.c
@@ -209,7 +209,7 @@ void ath11k_debugfs_sta_add_tx_stats(struct ath11k_sta *arsta,
 	tx_stats->ru_start = peer_stats->ru_start;
 	tx_stats->ru_tones = peer_stats->ru_tones;
 
-	if (peer_stats->mu_grpid <= MAX_MU_GROUP_ID &&
+	if (peer_stats->mu_grpid < MAX_MU_GROUP_ID &&
 	    peer_stats->ppdu_type != HTT_PPDU_STATS_PPDU_TYPE_SU) {
 		if (peer_stats->mu_grpid & (MAX_MU_GROUP_ID - 1))
 			tx_stats->mu_group[peer_stats->mu_grpid] =
-- 
2.34.1

